ngraph/core/reference/include/ngraph/runtime/reference/convolution_backprop_data.hpp

```diff
@@ -23,217 +19,224 @@ namespace ngraph
+                    extend_to_3D(params, input_shape, filters_shape);
+                }
 
-                    ACCUMULATION result = 0;
+                for (size_t i = 0; i < input_shape.size() - 2; ++i)
+                {
+                    if (input_shape[i + 2] > 1 || filters_shape[i + 2] > 1)
+                    {
+                        params.pads_begin[i] = filters_shape[i + 2] - params.pads_begin[i] - 1;
+                        params.pads_end[i] = filters_shape[i + 2] - params.pads_end[i] - 1;
+                    }
+                    else
+                    {
+                        params.pads_begin[i] = 0;
+                        params.pads_end[i] = 0;
+                    }
+                }
 
-                    CoordinateTransform::Iterator in_it = in_transform.begin();
-                    CoordinateTransform::Iterator filter_it = filter_transform.begin();
-                    CoordinateTransform::Iterator in_it_end = in_transform.end();
-                    CoordinateTransform::Iterator filter_it_end = filter_transform.end();
+                // convert output shape to 3D, contains only dimensions
+                Shape out_shape_3d{out_shape.begin() + 2, out_shape.end()};
 
-                    size_t in_channel_stride = row_major_strides(in_shape).at(in_channel_axis);
-                    size_t filter_in_channel_stride =
-                        row_major_strides(filter_shape).at(filter_in_channel_axis);
+                int out_shape_rank = out_shape.size() - 2;
+                if (out_shape_rank < 3)
+                {
+                    int missing_dims = 3 - out_shape_rank;
+                    out_shape_3d.insert(
+                        std::next(out_shape_3d.end(), -out_shape_rank), missing_dims, 1);
+                }
 
-                    while (in_it != in_it_end && filter_it != filter_it_end)
+                // modify params.pads_end when output_shape was provided in ctor in order to
+                // calculate expected number of output elements
+                for (size_t i = 0; i < out_shape_3d.size(); i++)
+                {
+                    if (out_shape_3d[i] > 1)
                     {
-                        const Coordinate& in_coord = *in_it;
-                        if (in_transform.has_source_coordinate(in_coord))
+                        // expected_dim = (in - 1)* strides + filter - 2*padding + out_padding
+                        // strides is already applied (through 0's extension in input)
+                        // padding = pads_begin + pads_end, formula below is using
+                        // params.pad_begin/params.pads_end:
+                        size_t expected_dim = (input_shape[i + 2] - 1) - filters_shape[i + 2] +
+                                              params.pads_begin[i] + params.pads_end[i] + 2 +
+                                              params.output_padding[i];
+                        if (out_shape_3d[i] != expected_dim)
```

Comment:
I think, it's useless condition
Because if `out_shape_3d[i] == expected_dim` then `out_shape_3d[i] - expected_dim = 0` and `params.pads_end[i] += 0`. It's ok

So I suggest:
`const size_t expected_dim = out_shape_3d[i] - ((input_shape[i + 2] - 1) - filters_shape[i + 2] +
                                              params.pads_begin[i] + params.pads_end[i] + 2 +
                                              params.output_padding[i]);`
`params.pads_end[i] += expected_dim;`

Reply:
You are right, if statement is redundant here. Thanks for this catch:)